src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalPane.java

```diff
@@ -284,7 +299,78 @@ public void terminateAllTerminals()
       terminals_.clear();
       creatingTerminal_ = false;
       activeTerminalToolbarButton_.setNoActiveTerminal();
-      setTerminalCaption("");
+      setTerminalTitle("");
+   }
+
+   @Override
+   public void renameTerminal()
+   {
+      final TerminalSession visibleTerminal = getSelectedTerminal();
+      if (visibleTerminal == null)
+      {
+         return;
+      }
+      final String origCaption = visibleTerminal.getCaption();
+
+      globalDisplay_.promptForText("Rename Terminal",
+            "Please enter the new terminal name:",
+            origCaption,
+            0,
+            origCaption.length(),
+            null,
+            new ProgressOperationWithInput<String>()
+            {
+               @Override
+               public void execute(final String newCaption,
+                                   final ProgressIndicator progress)
+               {
+                  progress.onProgress("Renaming terminal...");
```

Comment:
Nit: Consider using something like `DelayedProgressRequestCallback` here so that the user doesn't see the progress message unless it takes more than a few ms to execute.